src/zarr/core/metadata/v2.py

```diff
@@ -284,3 +284,19 @@ def parse_fill_value(fill_value: object, dtype: np.dtype[Any]) -> Any:
             raise ValueError(msg) from e
 
     return fill_value
+
+
+def default_fill_value(dtype: np.dtype[Any]) -> Any:
+    """
+    Get the default fill value for a type.
+
+    Notes
+    -----
+    This differs from :func:`parse_fill_value`, which parses a fill value
+    stored in the Array metadata into an in-memory value. This only gives
+    the default fill value for some type.
+
+    This is useful for reading Zarr V2 arrays, which allow the fill
+    value to be unspecified.
+    """
+    return dtype.type(0)
```

Comment:
What about string or object dtypes? Does 0 really make sense there? In #2036 I used `'' and b''` for strings and bytes.

Reply:
Here's the zarr v2 behavior

```
In [13]: zarr.create(shape=(1,), store={}, path="a", dtype="|U1")[:]
Out[13]: array([''], dtype='<U1')

In [14]: zarr.create(shape=(1,), store={}, path="a", dtype="|S1")[:]
Out[14]: array([b''], dtype='|S1')

In [15]: zarr.create(shape=(1,), store={}, path="a", dtype=str)[:]
Out[15]: array([0], dtype=object)
```

The integer `0` there for `str` dtype (which zarr-python 2.x v2 converted to an `object` dtype) doesn't really make sense.

Reply:
ff530c36e41fd725f5a4cdc5e7b35a92a9668e2b restores the 2.x behavior aside from `dtype=str`. In 2.x, that was somehow special cased to use object dtype. Right now on main, that returns a `<U`-dtype array, similar to NumPy:

```
In [1]: import numpy as np

In [2]: np.array([1, 2], dtype=str)
Out[2]: array(['1', '2'], dtype='<U1')
```

I think the treatment of str / object-dtype arrays is mostly orthogonal to this PR, but we might need to update the test I added if we want to change the behavior there.

Reply:
I seriously doubt that anyone is relying on that [weird] behavior. 😆